### PR TITLE
[SPARK-13422][SQL] Use HashedRelation instead of HashSet in Left Semi Joins

### PR DESCRIPTION
**Pull request by xguo27**
Use the HashedRelation which is a more optimized datastructure and reduce code complexity

### CODE REVIEW

**Comment by xguo27** [comment]
@hvanhovell Could you please advise whether this is the right fix? All Left Semi related tests passed, but I'm not sure what other impact there might be to remove HashSet related methods.

**Comment by hvanhovell** [comment]
ok to test

**Comment by xguo27** [comment]
Looks like the command did not trigger a test?

**Comment by hvanhovell** [comment]
It usually takes about 15 min before a test is triggered

**Comment by xguo27** [comment]
@hvanhovell  I see, sorry for my lack of patience. : )

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastLeftSemiJoinHash.scala [hunk 1]
```diff
@@ -45,18 +45,15 @@ case class BroadcastLeftSemiJoinHash(
...
     }.collect()
 
-    if (condition.isEmpty) {
-      val hashSet = buildKeyHashSet(input.toIterator)
-      val broadcastedRelation = sparkContext.broadcast(hashSet)
+    val hashRelation =
+      HashedRelation(input.toIterator, rightKeyGenerator, input.size)
+    val broadcastedRelation = sparkContext.broadcast(hashRelation)
 
+    if (condition.isEmpty) {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastLeftSemiJoinHash.scala]: I think it is safe to remove the entire block.

**Comment by hvanhovell** [comment]
@xguo27 Could you also take a look at the `hashSemiJoin(...)` function and see how this treats an empty condition?

**Comment by hvanhovell** [comment]
I don't think I have enough karma to trigger a build... @rxin could you trigger a build?

**Comment by xguo27** [comment]
@hvanhovell In hashSemiJoin() function, when condition is empty, the boundCondition always evaluates to true here:

https://github.com/apache/spark/blob/8f744fe3d931c2380613b8e5bafa1bb1fd292839/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashSemiJoin.scala#L42-L43

so the exists{...} part of these lines behaves as a No-Op.

https://github.com/apache/spark/blob/8f744fe3d931c2380613b8e5bafa1bb1fd292839/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashSemiJoin.scala#L87-L89

**Comment by hvanhovell** [comment]
@xguo27 how expensive is the no-op? I'd rather avoid a pointless iteration in a hot code section.

**Comment by hvanhovell** [comment]
@xguo27 you'll need rebase this. We just merged a PR that changes broadcast joins.

**Comment by xguo27** [comment]
@hvanhovell I just rebased with your new PR, do you mind reviewing again?

**Comment by hvanhovell** [comment]
LGTM - pending tests

**Comment by hvanhovell** [comment]
retest this please

**Comment by rxin** [comment]
Thanks - merging in master.

**Pull request closed without merging**